sympy/functions/combinatorial/factorials.py

```diff
@@ -1154,40 +1155,53 @@ def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
 
     def _eval_rewrite_as_factorial(self, n, k, **kwargs):
         from sympy import Pow
+        isint = lambda x: ask(Q.integer(x))
+        isneg = lambda x: ask(Q.negative(x))
+        isnneg = lambda x: ask(Q.nonnegative(x))
+        iszero = lambda x: ask(Q.zero(x))
 
-        if fuzzy_and([n.is_integer, k.is_integer]):
-            if n.is_negative:
-                if k.is_nonnegative:
+        if isnneg(n) and isnneg(k) or any(x is False for x in [isint(n), isint(k)]):
+            if iszero(n - k + 1):
+                return S.Zero
+            else:
+                return factorial(n)/(factorial(k)*factorial(n - k))
+        if fuzzy_and([isint(n), isint(k)]):
+            if isneg(n):
+                if isnneg(k):
                     return(Pow(-1, k)*factorial(k - n - 1)/(
                         factorial(k)*factorial(-n - 1)))
-                if k.is_negative:
+                if isneg(k):
                     return (Pow(-1, n - k)*factorial(-k - 1)/(
                         factorial(n - k)*factorial(-n - 1)))
-            elif n.is_nonnegative:
-                if k.is_negative:
+            elif isnneg(n):
+                if isneg(k):
                     return S.Zero
-                if k.is_nonnegative:
+                if isnneg(k):
                     return factorial(n)/(factorial(k)*factorial(n - k))
-        if fuzzy_and([n.is_integer, k.is_integer]) is False:
-            return factorial(n)/(factorial(k)*factorial(n - k))
 
     def _eval_rewrite_as_gamma(self, n, k, **kwargs):
         from sympy import gamma, Pow
+        isint = lambda x: ask(Q.integer(x))
```

Comment:
So the choice is new assumptions or keyword-flavored binomial? I prefer the keyword.

Reply:
I fear, I may not be well informed to make a decision here. 😅 
What gives us the most versatility and usability ?

Reply:
this works 😃, so I like it

Reply:
I haven't appreciated the use of the new assumptions much. This application is interesting.

You should probably use the `ask`-idiom in the other rewrite routines (like `rewrite_as_factorial`). Have you compared the times to rewrite to gamma for `(n,k)` and `(m n - k)` with `k < n` in the 2nd case? or old-style (n,k) vs new-style (n,k) rewrites?

Reply:
I guess the good news is that a lot of the tested functionality doesn't need the assumptions spelled out.

Reply:
> You should probably use the `ask`-idiom in the other rewrite routines (like `rewrite_as_factorial`).

Okay I'll do that.


> Have you compared the times to rewrite to gamma for `(n,k)` and `(m n - k)` with `k < n` in the 2nd case? or old-style (n,k) vs new-style (n,k) rewrites?

Nope, not explicity, but while running the tests I think I see that the new style is kinda faster.
I'll confirm that.

> I guess the good news is that a lot of the tested functionality doesn't need the assumptions spelled out.

Yeah, we fixed the assumptions for rewriting right, setting `nonnegative=True` is like restricting binomial to the `Knuth` definition. except cases where negative ints may spill over, thats where we need the correct assumptions.

`test_issue_8016` doesn't work in WolframAlpha cause of the <b>spilling over</b>.
 it works only when restricted to the F.C.K definition